File: clients/fsharp/Model/Action.fs
#nowarn "0058"

namespace SpbAiChamp
namespace SpbAiChamp.Model

open SpbAiChamp

/// Player's actions
type Action = {
    /// List of movement orders
    Moves: Model.MoveAction[];
    /// List of building orders
    Buildings: Model.BuildingAction[];
} with

    /// Write Action to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.Moves.Length
        this.Moves |> Array.iter (fun value ->
            value.writeTo writer )
        writer.Write this.Buildings.Length
        this.Buildings |> Array.iter (fun value ->
            value.writeTo writer )
        ()

    /// Read Action from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Moves = [|for _ in 1 .. reader.ReadInt32() do
                    yield Model.MoveAction.readFrom reader; |]
        Buildings = [|for _ in 1 .. reader.ReadInt32() do
                        yield Model.BuildingAction.readFrom reader; |]
    }

File: clients/rust/src/model/action.rs
use super::*;

/// TODO - Document
#[derive(Clone, Debug)]
pub struct Action {
    /// TODO - Document
    pub moves: Vec<model::MoveAction>,
    /// TODO - Document
    pub buildings: Vec<model::BuildingAction>,
}

impl trans::Trans for Action {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.moves.write_to(writer)?;
        self.buildings.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let moves: Vec<model::MoveAction> = trans::Trans::read_from(reader)?;
        let buildings: Vec<model::BuildingAction> = trans::Trans::read_from(reader)?;
        Ok(Self {
            moves,
            buildings,
        })
    }
}